--- GobiSerial/GobiSerial.c	2011-09-07 00:04:53.000000000 +0200
+++ GobiSerial.mod/GobiSerial.c	2012-07-30 15:23:43.274778696 +0200
@@ -153,6 +153,7 @@
 {
    { USB_DEVICE( 0x05c6, 0x920c ) },   // Gobi 3000 QDL device
    { USB_DEVICE( 0x05c6, 0x920d ) },   // Gobi 3000 Composite Device
+   { USB_DEVICE( 0x03f0, 0x371d ) },   // Gobi 3000 HP un2430 Device
    { }                               // Terminating entry
 };
 MODULE_DEVICE_TABLE( usb, GobiVIDPIDTable );
